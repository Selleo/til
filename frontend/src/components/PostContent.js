import React, { useMemo } from 'react'

import { Link, useLocation } from 'react-router-dom'
import { parseISO } from 'date-fns'

import { timeFormat } from '../utils'
import { useIsPostPublic } from '../utils/customHooks/useIsPostPublic'
import useUser from '../utils/customHooks/useUser'

import Avatar from './Avatar'
import CopyPostURL from './CopyURL'
import Markdown from './Markdown'
import PostCategories from './PostCategories'
import ReactionBar from './ReactionBar'
import TextBlock from './TextBlock'
import { Transition } from './Transition'
import UserPostMenu from '../authenticated/UserPostMenu'

const PostContent = ({ animationDelay, post, review, userMenu }) => {
  const { pathname } = useLocation()
  const user = useUser()

  const isPublic = useIsPostPublic(post.isPublic)
  const parsed = parseISO(post.createdAt)
  const date = timeFormat(parsed)

  const isPostOwner = useMemo(() => {
    if (user && post) {
      return user.uuid === post.author.uuid
    }
  }, [user, post])

  let title
  if (pathname === '/search') {
    title = <TextBlock value={post.title} />
  } else {
    title = post.title
  }

  const TitleLink = () => {
    return review ? (
      <span className="post__title">{title}</span>
    ) : (
      <Link className="post__title" to={`/posts/${post.id}-${post.slug}`}>
        {title}
      </Link>
    )
  }

  return (
    <Transition name="post-animation" delay={animationDelay}>
      <article
        className="post"
        style={{ transitionDelay: `${animationDelay}ms` }}
      >
        <div className="post__header">
          <div className="post__details">
            <Link
              to={isPostOwner ? `/profile` : `/authors/${post.author.userName}`}
              className="post__link user-avatar"
            >
              <Avatar imageUrl={post.author.image} background="light" />
            </Link>
            <div className="post__text-details">
              <Link
                to={
                  isPostOwner ? `/profile` : `/authors/${post.author.userName}`
                }
                className="post__link"
              >
                <div className="post__owner">
                  <span className="animation">
                    {post.author.firstName} {post.author.lastName}
                  </span>
                </div>
              </Link>
              <div className="post__date">{date}</div>
              <div className="post__is-public">{isPublic}</div>
            </div>
          </div>

          {!review && <CopyPostURL postId={post.id} />}
        </div>
        <div>
          <TitleLink />
        </div>
        <div className="post__body">
          <Markdown source={post.body} />
        </div>
        <div className="post__footer">
          <PostCategories categories={post.categories} />
          {!review && <ReactionBar post={post} />}
        </div>
        {(userMenu || isPostOwner) && (
          <UserPostMenu post={post} userMenu={userMenu} isOnReview={review} />
        )}
      </article>
    </Transition>
  )
}

export default PostContent
