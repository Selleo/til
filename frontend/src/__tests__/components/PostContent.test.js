import React from 'react'
import { prettyDOM, render, screen } from '@testing-library/react'
import PostContent from '../../components/PostContent'
import renderWithStoreAndRouter from '../../tests/utils/renderWithStoreAndRouter'

describe('PostContent', () => {
  const postMock = {
    id: 'test-id',
    isPulic: true,
    slug: 'test-slug',
    createdAt: '2022-04-21T09:39:26',
    author: {
      userName: 'Tenteges',
      firstName: 'Tenteges',
      lastName: 'Wihajster',
      image: 'http://placekitten.com/200/',
    },
    reactions: [
      {
        post_id: 'test-id',
        type: 'like',
        user_uuid: '1',
      },
    ],
  }

  it('should render properly', () => {
    const { container } = renderWithStoreAndRouter(
      <PostContent post={postMock} />,
      { route: '/search' }
    )
    expect(container).toMatchSnapshot()
  })

  it('should render with userMenu', () => {
    renderWithStoreAndRouter(<PostContent post={postMock} userMenu />, {
      route: '/search',
    })
    expect(screen.getByText('Edit')).toBeInTheDocument()
    expect(screen.getByText('Delete')).toBeInTheDocument()
  })
})
