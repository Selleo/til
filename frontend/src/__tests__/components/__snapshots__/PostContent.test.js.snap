// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostContent should render properly 1`] = `
<div>
  <article
    class="post post-animation-appear post-animation-appear-active"
    style="transition-delay: undefinedms;"
  >
    <div
      class="post__header"
    >
      <div
        class="post__details"
      >
        <a
          class="post__link user-avatar"
          href="/authors/Tenteges"
        >
          <img
            alt="author-img"
            class="user__image"
            src="http://placekitten.com/200/"
          />
        </a>
        <div
          class="post__text-details"
        >
          <a
            class="post__link"
            href="/authors/Tenteges"
          >
            <div
              class="post__owner"
            >
              <span
                class="animation"
              >
                Tenteges
                 
                Wihajster
              </span>
            </div>
          </a>
          <div
            class="post__date"
          >
             21 Apr 09:39
          </div>
          <div
            class="post__is-public"
          />
        </div>
      </div>
      <div
        aria-describedby="tippy-tooltip-1"
        class="ToolTip"
        data-tooltipped=""
        style="display: inline;"
      >
        <button
          class="post__share-button"
        >
          <svg>
            share.svg
          </svg>
          <span
            class="post__share-text"
          >
            Share
          </span>
        </button>
      </div>
    </div>
    <div>
      <span
        class="post__title"
      />
    </div>
    <div
      class="post__body"
    />
    <div
      class="post__footer"
    >
      <div
        class="post__reactions"
      >
        <div
          class="post__single-reaction"
        >
          <svg
            class="like"
          >
            like.svg
          </svg>
          <div
            class="reaction-number"
          >
            0
          </div>
        </div>
        <div
          class="post__single-reaction"
        >
          <svg
            class="funny"
          >
            funny.svg
          </svg>
          <div
            class="reaction-number"
          >
            0
          </div>
        </div>
        <div
          class="post__single-reaction"
        >
          <svg
            class="love"
          >
            love.svg
          </svg>
          <div
            class="reaction-number"
          >
            0
          </div>
        </div>
        <div
          class="post__single-reaction"
        >
          <svg
            class="surprised"
          >
            surprised.svg
          </svg>
          <div
            class="reaction-number"
          >
            0
          </div>
        </div>
      </div>
    </div>
  </article>
</div>
`;
